### PR TITLE
add documents of group convolution tensor layout

### PR DESCRIPTION
**Pull request by asleepzzz**
we create a document to record the layout of group convolution

### CODE REVIEW

**Review by whchung — CHANGES_REQUESTED** [review]
The purpose the document is not stated in the beginning.

**Comment by whchung** [comment]
@asleepzzz Ask any random MIOpen developer who knows NOTHING about MLIR and ask for his/her review of this document.

**Comment by whchung** [comment]
Also the file name appears to be pretty generic. Does it mean the memo covers everything wrt "tensor layout"? Or does it only cover specifics wrt "tensor layouts used in group convolutions"?

**Comment by whchung** [comment]
markdown texts shall be better stored under `docs` directory.
And maybe create a `MIOpen` directory under that.

**Comment by whchung** [comment]
The subject of this PR is also better be changed to reflect the document is created specifically for "group convolution".

**Comment by whchung** [comment]
@carlushuang / @shaojiewang Without knowing anything in MLIR, please help review the document in this PR. Please see if you guys can understand what @asleepzzz and @ltqin are trying to convey. And help suggest what needs to be amended in the memo.

File: mlir/docs/MIOpen/tensor_layout.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+2,2,2,1
+2,2,2,1
+[[[[1]],[[3]]],[[[5]],[[7]]]],[[[[9]],[[11]]],[[[13]],[[15]]]]
+[[[[2]],[[4]]],[[[6]],[[8]]]],[[[[10]],[[12]]],[[[14]],[[16]]]]
+<br/>
+#####user format is [YXGCK], not final format
+weights = tf.get_variable('weights', shape=[filter_height,     
+                                                    filter_width,
+                                                    input_channels/groups,
+                                                    num_filters])
```
Comment (carlushuang) [inline mlir/docs/MIOpen/tensor_layout.md:33]: `[YXGCK]` and here `tf.get_variable()` create a 4d tensor? where to parse in the `groups` information to the `weights` variable?

File: mlir/docs/MIOpen/tensor_layout.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+##tensorflow
+###NHWC
+<br/>
+<br/>
+####When you want to add group in tensorflow,
+####you split on dim of c or k , and tf.concat outputs 
+<br/>
+#####axis=3 split on c dim ,input layout become [N H W G C]
+input_groups = tf.split(axis=3, num_or_size_splits=groups, value=x)
```
Comment (carlushuang) [inline mlir/docs/MIOpen/tensor_layout.md:12]: suggest add add some description of what this function is doing, e.g.:

this function will split `[N, H, W, C]` tensor `x` along the 3rd dimension, result in `groups` number of tensors with `[N, H, W, C/groups]` each

File: mlir/docs/MIOpen/tensor_layout.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+weights = tf.get_variable('weights', shape=[filter_height,     
+                                                    filter_width,
+                                                    input_channels/groups,
+                                                    num_filters])
+<br/>
+#####but tensorflow will  transform filter from YXCK to KYXC:
+https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/kernels/conv_ops.cc#L1040-L1048
+<br/>
+#####axis=3 split on k dim before transform, after transform , weights layout become [GKYXC]
+weight_groups = tf.split(axis=3, num_or_size_splits=groups,value=weights)
```
Comment (carlushuang) [inline mlir/docs/MIOpen/tensor_layout.md:41]: `weights = tf.get_variable(` and `weight_groups = tf.split(` need to be used together to get the TRUE group weights right? suggest to put two python statement together, or other might be confusing.

File: mlir/docs/MIOpen/tensor_layout.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+<br/>
+from: https://www.tensorflow.org/api_docs/python/tf/nn/conv2d
+results is NHWK
+#####finally we concat at dim k ,the layout is [NHWGK]
+conv = tf.concat(axis=3, values=output_groups)
```
Comment (carlushuang) [inline mlir/docs/MIOpen/tensor_layout.md:48]: here a suddenly appear variable `output_groups`, this is the result of input convolve weight right? suggest add code or description about the whole process.

File: mlir/docs/MIOpen/tensor_layout.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+conv_nhwc = nn.Conv2d(C, K, 3, 1, 1).to("cuda:0", memory_format=torch.channels_last)
+we can dump shape like conv_nhwc.shape or a= nn.Conv2d a.weight.shape
+then we can judge G in which dim because if G=3 C=6 input=[N 6 H W] ,we can know dim(G) is dim(C)-1
+####input:[NHWGC]
+####filter:[GKCYX]
+####output:[NHWGK]
+<br/>
+<br/>
+if filter:YXCK, tensorflow will transform it and pytorch not use it, we can set GYXCK,
+but if it from tensorflow, don't use this format even user set it
```
Comment (carlushuang) [inline mlir/docs/MIOpen/tensor_layout.md:90]: what do you mean by 'if it from tensorflow, don't use this format even user set it', you mean even if `YXCK` set by user, don't use it by MLIR? From previous sentence, tensorflow will transform it (into KYXC), MLIR will see KYXC, why not use it? suggest more describe what's going on here

**Review by carlushuang — COMMENTED** [review]
some sentence is confusing for me maybe not familiar with MLIR

File: mlir/docs/MIOpen/tensor_layout.md (hunk 1 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:33]: groups is not real dim ,this sentence just showed original dim before split

File: mlir/docs/MIOpen/tensor_layout.md (hunk 3 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:41]: unfortunately, it transform by tf before split, I know some guys may be confused,
but tf  don't just split  the tensor

File: mlir/docs/MIOpen/tensor_layout.md (hunk 5 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:90]: if your tensor is from tf , it will become KYXC , but if directly pass to MLIR, YXCK

File: mlir/docs/MIOpen/tensor_layout.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+### NHWC
+<br/>
+<br/>
+#### When you want to add group in tensorflow,
+#### you split on dim of c or k , and tf.concat outputs 
+<br/>
+##### axis=3 split on c dim ,input layout become [N H W G C]
+input_groups = tf.split(axis=3, num_or_size_splits=groups, value=x)
+this function will split [N, H, W, C] tensor x along the 3rd dimension, result in groups number of tensors 
+with [N, H, W, C/groups] each ,so it's equivalent to [N H W G C]
```
Comment (shaojiewang) [inline mlir/docs/MIOpen/tensor_layout.md:14]: What does `value=x` and `tensor x` means here?

File: mlir/docs/MIOpen/tensor_layout.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+if nhwk is 
+2,2,2,2
+[[[[1],[2]],[[3],[4]]],[[[5],[6]],[[7],[8]]]],[[[[9],[10]],[[11],[12]]],[[[13],[14]],[[15],[16]]]]
+=>
+<br/>
+after split to 2 groups :
+2,2,2,1
+2,2,2,1
+[[[[1]],[[3]]],[[[5]],[[7]]]],[[[[9]],[[11]]],[[[13]],[[15]]]]
+[[[[2]],[[4]]],[[[6]],[[8]]]],[[[[10]],[[12]]],[[[14]],[[16]]]]
```
Comment (shaojiewang) [inline mlir/docs/MIOpen/tensor_layout.md:26]: I think that is more like a [GNHWK]?

File: mlir/docs/MIOpen/tensor_layout.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+#### filter:[GKCYX]
+<br/>
+### NHWC
+<br/>
+<br/>
+we can use channels_last to force pytorch use NHWC
+b = torch.empty(N, C, H, W, device="cuda:0", memory_format=torch.channels_last)
+conv_nhwc = nn.Conv2d(C, K, 3, 1, 1).to("cuda:0", memory_format=torch.channels_last)
+we can dump shape like conv_nhwc.shape or a= nn.Conv2d a.weight.shape
+then we can judge G in which dim because if G=3 C=6 input=[N 6 H W] ,we can know dim(G) is dim(C)-1
```
Comment (shaojiewang) [inline mlir/docs/MIOpen/tensor_layout.md:82]: I do not fully understand this line. Why `dim(G)=dim(C) - 1`?

File: mlir/docs/MIOpen/tensor_layout.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#### input:[NHWGC]
+#### filter:[GKCYX]
+#### output:[NHWGK]
+<br/>
+<br/>
+if filter:YXCK, tensorflow will transform it and pytorch not use it, we can set GYXCK,
+but if it from tensorflow, don't use this format even user set it
+from tensorflow: YXCK=> KCYX or KYXC
```
Comment (shaojiewang) [inline mlir/docs/MIOpen/tensor_layout.md:91]: `Groups` dimension will be added to which axis by tf?

**Review by shaojiewang — COMMENTED** [review]
I almost understand this document. I have some questions in detail.

File: mlir/docs/MIOpen/tensor_layout.md (hunk 6 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:14]: it's tf api ,add description,
tensor x is a tensor..... , we use it as an example

File: mlir/docs/MIOpen/tensor_layout.md (hunk 7 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:26]: there are comments
after split to 2 groups :

but not [GNHWK], because split on last dim

File: mlir/docs/MIOpen/tensor_layout.md (hunk 8 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:82]: if you put G to other place ,you can't get N 6 H W

File: mlir/docs/MIOpen/tensor_layout.md (hunk 9 above)
Comment (asleepzzz) [reply mlir/docs/MIOpen/tensor_layout.md:91]: No, G dim is not real dim ,but tf transform the total format,
I just list if we want to use G ,what will we get
G is a fake dim , you can use it to do more tensor transform, 
but it's not in tf tensor

**Comment by whchung** [comment]
@asleepzzz please resolve discussions which have been answered, and reply to those which have not

**Comment by asleepzzz** [comment]
> @asleepzzz please resolve discussions which have been answered, and reply to those which have not

done

**Comment by whchung** [comment]
jenkins: retest this please.

**Review by whchung — APPROVED** [review]
